instrumentation/dubbo/src/main/java/brave/dubbo/TracingFilter.java

```diff
@@ -139,39 +133,18 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
         deferFinish = true;
         // NOTE: We don't currently instrument CompletableFuture, so callbacks will not see the
         // invocation context unless they use an executor instrumented by CurrentTraceContext
-        ((CompletableFuture<?>) future).whenComplete((v, t) -> {
-          if (t != null) onError(t, span);
-          span.finish();
-        });
+        // If we later instrument this, take care to use the correct context depending on RPC kind!
+        ((CompletableFuture<?>) future)
+          .whenComplete(FinishSpan.create(this, request, result, span));
       }
       return result;
     } catch (Throwable e) {
       propagateIfFatal(e);
       error = e;
       throw e;
     } finally {
-      if (error != null) onError(error, span);
-      if (!deferFinish) span.finish();
+      if (!deferFinish) FinishSpan.finish(this, request, result, error, span);
```

Comment:
Not really something for this PR, but just wondering, does this mean the timestamp of the span will be bogus?

Reply:
we usually have synchronous methods finish once they return, whether that's a synchronous exception or result.

Reply:
rewrote the code around it a little.